### PR TITLE
feat: special magic level

### PR DESCRIPTION
**Pull request by ArturKnopik**
16:06 You see a rainbow shield (Def:30, energy magic level +20, fire magic level +3, healing magic level +10).
It can only be wielded properly by knights of level 100 or higher.
It weighs 69.00 oz.
Strange elemental magic flows over this shield.
specialSkillMlvl
{"magiclevelenergy", ITEM_PARSE_MAGICLEVELENERGY},
{"magiclevelfire", ITEM_PARSE_MAGICLEVELFIRE},
{"magiclevelpoison", ITEM_PARSE_MAGICLEVELPOISON},
{"magiclevelearth", ITEM_PARSE_MAGICLEVELPOISON},
{"magiclevelice", ITEM_PARSE_MAGICLEVELICE},
{"magiclevelholy", ITEM_PARSE_MAGICLEVELHOLY},
{"magicleveldeath", ITEM_PARSE_MAGICLEVELDEATH},
{"magiclevellifedrain", ITEM_PARSE_MAGICLEVELLIFEDRAIN},
{"magiclevelmanadrain", ITEM_PARSE_MAGICLEVELMANADRAIN},
{"magicleveldrown", ITEM_PARSE_MAGICLEVELDROWN},
{"magiclevelphysical", ITEM_PARSE_MAGICLEVELPHYSICAL},
{"magiclevelhealing", ITEM_PARSE_MAGICLEVELHEALING},
{"magiclevelundefined", ITEM_PARSE_MAGICLEVELUNDEFINED},

### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### CODE REVIEW

**Comment by Zbizu** [comment]
I believe that they should also be pushed to lua (in itemType:getAbilities())

File: src/items.h [hunk 1]
```diff
@@ -188,7 +200,7 @@ struct Abilities {
 	//extra skill modifiers
 	std::array<int32_t, SKILL_LAST + 1> skills = {0};
 	std::array<int32_t, SPECIALSKILL_LAST + 1> specialSkills = {0};
-
+	std::array<int16_t, COMBAT_COUNT> specialMagicLevelSkill = { 0 };
```
Comment (EPuncker) [inline src/items.h:203]:
```suggestion
	std::array<int16_t, COMBAT_COUNT> specialMagicLevelSkill = {0};
```

File: src/player.h [hunk 2]
```diff
@@ -1283,6 +1294,7 @@ class Player final : public Creature, public Cylinder
 		int32_t varSkills[SKILL_LAST + 1] = {};
 		int32_t varSpecialSkills[SPECIALSKILL_LAST + 1] = {};
 		int32_t varStats[STAT_LAST + 1] = {};
+		std::array<int16_t, COMBAT_COUNT> specialMagicLevelSkill = { 0 };
```
Comment (EPuncker) [inline src/player.h:1297]:
```suggestion
		std::array<int16_t, COMBAT_COUNT> specialMagicLevelSkill = {0};
```

File: src/item.cpp [hunk 3]
```diff
@@ -1084,7 +1084,44 @@ std::string Item::getDescription(const ItemType& it, int32_t lookDistance,
...
+					}
+
+					if (tmp) {
+						tmp = false;
+
+						if (begin) {
+							begin = false;
+							s << " (";
+						}
+						else {
```
Comment (EPuncker) [inline src/item.cpp:1112]: `} else {` same with others, other than that, this PR is ready to be merged

**Review by EPuncker — CHANGES_REQUESTED** [review]

File: src/protocolgame.cpp [hunk 4]
```diff
@@ -2260,6 +2260,20 @@ void ProtocolGame::sendMarketDetail(uint16_t itemId)
...
 
+		for (size_t i = 0; i < COMBAT_COUNT; ++i) {
+			if (it.abilities->specialMagicLevelSkill[i] == 0) {
+				continue;
+			}
+
+			if (separator) {
+				ss << ", ";
+			}
+			else {
```
Comment (omarcopires) [inline src/protocolgame.cpp:2271]: code style

File: src/item.cpp [hunk 5]
```diff
@@ -1227,7 +1264,44 @@ std::string Item::getDescription(const ItemType& it, int32_t lookDistance,
...
+							s << ", ";
+						}
+					}
+					else {
```
Comment (omarcopires) [inline src/item.cpp:1296]: code style

File: src/item.cpp [hunk 6]
```diff
@@ -1227,7 +1264,44 @@ std::string Item::getDescription(const ItemType& it, int32_t lookDistance,
...
+					}
+
+					if (tmp) {
+						tmp = false;
+
+						if (begin) {
+							begin = false;
+							s << " (";
+						}
+						else {
```
Comment (omarcopires) [inline src/item.cpp:1292]: code style

File: src/item.cpp [hunk 7]
```diff
@@ -1084,7 +1084,44 @@ std::string Item::getDescription(const ItemType& it, int32_t lookDistance,
...
+							s << ", ";
+						}
+					}
+					else {
```
Comment (omarcopires) [inline src/item.cpp:1116]: code style

File: src/item.cpp (hunk 3 above)
Comment (omarcopires) [inline src/item.cpp:1112]: code style

**Comment by nekiro** [comment]
What healing magic level stands for? It boosts magic level or the formula that decides how much you heal from magic level?

File: src/item.cpp (hunk 5 above)
Comment (ArturKnopik) [reply src/item.cpp:1296]: fixed

File: src/item.cpp (hunk 6 above)
Comment (ArturKnopik) [reply src/item.cpp:1292]: fixed

File: src/item.cpp (hunk 7 above)
Comment (ArturKnopik) [reply src/item.cpp:1116]: fixed

Comment (ArturKnopik) [reply src/item.cpp:1116]: fixed

File: src/item.cpp (hunk 3 above)
Comment (ArturKnopik) [reply src/item.cpp:1112]: fixed

File: src/protocolgame.cpp (hunk 4 above)
Comment (ArturKnopik) [reply src/protocolgame.cpp:2271]: fixed

**Review by EPuncker — APPROVED** [review]
tested and working :)

**Comment by omarcopires** [comment]
This feature ends a task from issue #3758
`New types of items modifiers were be implemented: Specialised Magic Levels, Cleave, Magic Shield Capacity, Perfect Shot (aka Range Bonus), Damage Reflection [12.70.10953]`

**Pull request merged**